core/domain/exp_domain.py

```diff
@@ -1344,6 +1357,7 @@ def update_title(self, title):
             title: str. The exploration title to set.
         """
         self.title = title
+        proto_size_is_stale = True
```

Comment:
This is OK in the interim, but see if you can find a way to have Python do something when a member variable is updated. @vojtechjelinek might be able to help you or offer suggestions. My concern with the current approach is that if someone adds another update method, or modifies exploration.XXX somewhere else in the code, this won't cause the stale flag to be triggered. So I think finding a more foolproof method would be better.

Reply:
From my understanding, if stale flag to be triggered at `someone adds another update method, or modifies exploration.XXX` then again this cause the same problem of recomputation of size!! isn't it?

We need some exit points so that only at the exit does the size calculation or places like when someone wants `from_dict`, `to_dict`, saving exp, etc.  At entry points, we can have 0 as the `proto_size_in_bytes` as the discussion over email.

Reply:
No -- the idea here behind the caching is to do the recomputation lazily, i.e. only when some other method tries to _access_ the proto size. There is no tracking of entry or exit points because these may change in the future. Note that 0 is for the initial value on initialization of the domain object, _and that value is considered stale anyway_.

Basically, this approach makes things context independent. At any time the domain object is accessed, either the computed value is stale or it is not. If it is stale (which is the case at the beginning too), then the size is computed and cached for future reads. If it is not stale then the last-cached size is read as-is. So please don't write the logic based on any assumptions about entry/exit points (though we might still want to do an additional recomputation at known "exit points" anyway just to be safe).

Given the above, I don't understand the concern in your first paragraph. Could you re-explain, perhaps? Basically, we want a general way to listen to any update methods so that we can ensure that the previously-computed size is flagged as stale at the right times. The concern with manually adding this to each update method is that someone who adds a new update method in the future may forget to add this call.

Reply:
I did a bit of research and the method you'll want to look into is `__setattr__`. This code snippet will show you how it works.

You can add some code to `__setattr__` to handle the 'stale' property any time a member property of the class is edited. Make sure you only consider the relevant properties (i.e. you don't need to update the stale value if it's the proto_size attribute of the class that's being changed).

```
class A(object):

    def __init__(self, value):
        self.x = value

    def __setattr__(self, aa, new_value):
        print('Setting property %s to %s' % (aa, new_value))
        super(A, self).__setattr__(aa, new_value)


obj = A(123)
obj.x = 456
print(obj.x)
```

Reply:
I agree with Sean's solution. But just a note about 'only consider the relevant properties', I think we need to find a way how to add these relevant properties in the future when we add some new properties to exploration. I think the best approach here would be a denylist that specifies for which properties not to update the `proto_size`.

Reply:
When does the `proto_size_in_bytes` change or update?

Reply:
Looks like fixed 🎉  Sample code snippet. 

```py
# Online Python compiler (interpreter) to run Python online.
# Write Python 3 code in this online editor and run it.
class A(object):

    def __init__(self):
        self.proto_size_in_bytes = None
        self._cached_android_proto_size_in_bytes = 0
        self.proto_size_is_stale = True
        self.name = 'Akshay'

    def __setattr__(self, aa, new_value):
        print('Setting property %s to %s' % (aa, new_value))
        super(A, self).__setattr__(aa, new_value)
        if aa == 'proto_size_is_stale':
            self.is_stale

    @property
    def is_stale(self):
        if self.proto_size_is_stale:
            self._cached_android_proto_size_in_bytes = 7756
            self.proto_size_is_stale = False
            self.proto_size_in_bytes = self._cached_android_proto_size_in_bytes

        return self._cached_android_proto_size_in_bytes


obj = A()
print(obj._cached_android_proto_size_in_bytes)
print(obj.proto_size_is_stale)
print(obj.proto_size_in_bytes)
```

Output
```console
Setting property proto_size_in_bytes to None
Setting property _cached_android_proto_size_in_bytes to 0
Setting property proto_size_is_stale to True
Setting property _cached_android_proto_size_in_bytes to 7756
Setting property proto_size_is_stale to False
Setting property proto_size_in_bytes to 7756
Setting property name to Akshay
7756
False
7756
```

Reply:
Added `__setattr__ `, but a bit confused with functions which are calling from outside. For example `save_new_exploration`, here we get the `proto_size_in_bytes` None, so the `__setattr__ ` didn't trigger when `save_new_valid_exploration` from `test_utils` was creating the exploration.

Reply:
Could you please explain the issue you're talking about more clearly? I don't understand what you mean by "here we get the proto_size_in_bytes None" and I'm afraid I don't follow your explanation of _why_ `__setattr__` didn't trigger.

Reply:
Thanks for the clearer explanation, I understand now. I think you'll need to figure out why, in your implementation, `__setattr__` isn't triggering when those two lines are run -- it should.

Reply:
It's working now.

Reply:
> Make sure you only consider the relevant properties (i.e. you don't need to update the stale value if it's the proto_size attribute of the class that's being changed).

Need a review on this part, is the current setattr function is correct or not? 
What are the revant properties? Are you talking about the properties which are used is proto size calculation only?